### ISSUE TITLE
[New Concept Exercise] : classes

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#2349]
This issue describes how to implement the `classes` concept exercise for the python track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [Contributing to Exercism](https://exercism.org/contributing) |  [Help](https://exercism.org/docs/building/product)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/tasks)
- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Exercise Specifications](https://exercism.org/docs/building/product/concept-exercises)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)  

## Goal

This concept exercise is meant to teach an understanding/use of `classes`  in Python. 

## Learning objectives

- understand what a `class` represents in Python
- how to declare (_create_) a `class`
- understanding that instantiating a `class` creates an `object`
- understand the difference between `class` `properties` vs `instance` `attributes`
- understanding that bound functions of a class are `methods` 
- the uses of `self`  to denote `method`
- creating/using `__init__()` to customize object creation
- using `pass` as a placeholder for class methods

## Out of scope

- `class-customiation`, including `@classmethod` & `@staticmethod`
-  `@property` decorator, getters, and  setters 
-  class members & non-public methods
- `class-inheritance`, `multiple-inheritance`, `__super()__`, class mixins
- `class-composition`
- `dataclasses`
- performance considerations

## Concepts

-  `attributes`
- `classes`
- `methods`
- `objects`
- `OOP`
-  `properties`

  
## Prerequisites

These are the concepts/concept exercises the student needs to complete/understand before solving this concept exercise.  

- `basics`
- `bools`
- `comparisons`
- `dicts`
- `iteration`
- `lists`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `tuples`

  

## Resources to refer to

- [classes (Python tutorial)](https://docs.python.org/3/tutorial/classes.html)
- [Python Data Model - Python Docs](https://docs.python.org/3/reference/datamodel.html#the-standard-type-hierarchy)
- [Real Python: Object-Oriented Programming in Python 3](https://realpython.com/python3-object-oriented-programming/)
- [DO: How To Construct Classes & Objects in Python 3](https://www.digitalocean.com/community/tutorials/how-to-construct-classes-and-define-objects-in-python-3)

* ### Hints

  For more information on writing hints see [hints](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-docshintsmd)

  * You can refer to one or more of the resources linked above, or analogous resources from a trusted source.  We prefer using links within the  [Python Docs](https://docs.python.org/3/) as the primary go-to, but other resources listed above are also good.  Please try to avoid paid or subscription-based links if possible.

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)

  *  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  *  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

## Concept Description

Please see the following for more details on these files:  [concepts](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson) & [concept exercises](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md)

* ### Concept `about.md`

  **Concept file/issue**: There is currently no issue or files for the concept.  **They are TBD.**

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-introductionmd)

  * This can also be a summary/paraphrase of the  document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the concept exercise.   It should contain a good summation of the concept, but not go into lots of detail.

* ### Exercise `introduction.md`

  For more information, see [Exercise `introduction.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-docsintroductionmd)

  * This should also summarize/paraphrase  the above document, but with  enough information and examples  for the student to complete the tasks outlined in this concept exercise.

  

## Test-runner

No changes required to the [Python Test Runner](https://github.com/exercism/python-test-runner) at this time.  

## Representer

No changes required to the [Python Representer](https://github.com/exercism/python-representer/) at this time.  

## Analyzer

No changes required to the [Python Analyzer](https://github.com/exercism/python-analyzer/) at this time.  

## Exercise Metadata - Track

For more information on concept exercises and formatting for the Python track `config.json` , please see [concept exercise metadata](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#metadata).  The track `config.json` file can be found in the root of the Python repo.

You can use the below for the exercise **UUID**.  You can also generate a new  one via [exercism configlet](https://github.com/exercism/configlet), [uuidgenerator.net](https://www.uuidgenerator.net/version4), or any other favorite method.  The UUID must be a valid  [**V4 UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier).

- **Exercise UUID** : `3550ec07-f6c6-48bd-b2b4-086e75faf9e7`
- **concepts** should be filled in from the Concepts section in this issue
- **prerequisites** should be filled in from the Prerequisites section in this issue  

## Exercise Metadata Files Under  `.meta/config.json`

For more information on exercise  `.meta/`  files and formatting, see [concept exercise metadata files](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#metadata-files)

- `.meta/config.json` - see [this link](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-metaconfigjson) for the fields and formatting of this file.
- `.meta/design.md` - see [this link](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-metadesignmd) for the formatting of this file.  Please use the **Goal**, **Learning Objectives**,**Concepts**, **Prerequisites** and , **Out of Scope**  sections  from this issue.

  

## Implementation Notes

Code in the `.meta/examplar.py` file should **only use syntax & concepts introduced in this exercise or one of its prerequisite exercises.**
Please **do not use** comprehensions, generator expressions, or other syntax not previously covered.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.

In General, tests should be written using `unittest.TestCase` and the test file should be named `<EXERCISE-NAME>_test.py`.

While we do use [PyTest](https://docs.pytest.org/en/stable/) as our test runner and for some implementation tests, please check with a maintainer before using  a PyTest test method, fixture,  or feature.

Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue, or contact one of the maintainers on our Slack channel.

**Comment by PaulT89** [comment]
I'd like to take a shot at this.  I'm thinking of having the user create an Enemy class for an otherwise unspecified game.  Not entirely sure how to demonstrate object creation though.  Maybe I'll have some standalone function where the user has to create a list of 3 different enemies, given some data passed to that function, and then return the list.

**Comment by PaulT89** [comment]
@BethanyG Okay, I've completed an exemplar file, a solution file, and a test file.  It's getting late, so I think I'll start on the documentation tomorrow.  I suspect that I may need help with everything else though (metadata files and whatnot).

**Comment by BethanyG** [comment]
@PaulT89  -- Late logging in today, so I missed you.  😞   **Delighted** you'd like to take this on!  I am happy to do the concept write-ups and other metadata stuff, if that helps -- and I look forward to seeing what you've come up with!  😄 🌟

**Comment by PaulT89** [comment]
@BethanyG What exactly should I do with the files that I have so far?  Do I just start making pull requests one by one (assuming there isn't a way to upload the files all at once), or do you want to review them first (i.e. before a deluge of PRs) or something else?  Also, I'm only vaguely acquainted with Slack, so I was wondering if I needed access to a particular channel to contact you.

**Comment by PaulT89** [comment]
I've created hints.md and instructions.md (locally), so I just need to write the introduction.  I suspect it will be simultaneously quite long and quite hand-wavy.  i.e. "Welcome to OOP - something something inheritance - but ignore that for now; something something polymorphism - fun word, ignore that too; something something, encapsulation - never mind that; here we go - classes, methods, and variables.  What's that?  How is this any different to what we were doing before with functions?  Well, you can do all sorts of cool stuff with OOP... like inheritance, polymorphism and encapsulation."  At least, this was my introduction to OOP.  It took a while for me to get it. 😜  The hard bit is going to be trying to hint in the introduction that there's more under the hood of classes and OOP in general, while still trying to maintain the student's attention on the fundamentals - fundamentals that don't look like they're offering anything new.

**Comment by BethanyG** [comment]
@PaulT89  -- if you are already a member of/in our Slack channel, you can DM me by selecting the `+` sign next to `Direct Messages` in the left-hand menu:

<img src="https://user-images.githubusercontent.com/5923094/150676165-dd748046-d771-4e71-98d2-d950429d1f53.png" width="400">

 You can also ask questions in the #discussions channel or #maintaining-python or #track-python.  If you are not a member, you'll need an invite.

As for the PR/files -- You can do a single PR for the exercise -- I typically (_after making a fork in GH and cloning the repo locally_) create a branch locally named after the exercise I am adding (_so -- something like `character-class-concept-exercise`_), then create the exercise directory, the sub-directories, and then add in the files and commit them.  Then I push to my GH fork and PR from there.  I typically mark anything that is still in process a `draft` -- but that is up to you, and how you like to work.  Here is an example PR for a merged concept exercise.  I think this one also includes the concept write-up :  [cater-waiter](https://github.com/exercism/python/pull/2485)

 [This spec](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md) has a file tree that shows how the concept exercise is laid out.  You can also take a look at [Guido's Gorgeous Lasagna](https://github.com/exercism/python/tree/main/exercises/concept/guidos-gorgeous-lasagna) in this repo to see an example of an exercise directory with its files.  You can also take a look at our [contributing docs](https://github.com/exercism/python/blob/main/CONTRIBUTING.md) for additional info and links.

If you are not comfortable with the git command line (or just want an alternative), GH has a decent desktop app that can help ease the circus of forking, branching and merging.  You can download it here:  https://desktop.github.com/

I need to take a bit of a nap, but should be back online in about 4-6 hours.  😄   Thanks again for your patience, and your willingness to create an exercise!

**Comment by PaulT89** [comment]
Okay thanks.  It's about 11pm here so I'll be going to bed soon anyway.  I'll look at this in the morning. 😄

**Comment by BethanyG** [comment]
I'll need an email to send an invite to (_if you don't want to note it publicly here, you can email it to me at bethany(at)pyladies(dot)com_), and I can send you an invite. Or you can also join through `exericsm.org`, if you sign up to be a mentor.  😄  

I ... feel you on the OOP/class thing.  And we may need to break it out into multiple docs.  _It's ok!! We'll get there together -- its a big-ass topic, so I expect this to have lots of iterations._  🌟

**Comment by PaulT89** [comment]
Okay, it's in the exercism/python/exercises/concept as 'ellens-alien-game'.  I've given the introduction and and design files the suffix -draft (as there's nothing in them yet).  

Other things: 
- While the test file works, and is verbose in its error messages, I'm not entirely sure I've laid it out in a way that would make perfect sense for a concept exercise.  I'll need feedback on whether or not I need to rework it.
- I'm not sure if the incrementing of a class variable (total_aliens_created) is too advanced for a beginner to wrap their head around, but I thought it was still technically within scope, and would at least show them something they (probably) couldn't do without classes.
- I already listed you as an author in the config file - you've already given me so much help 😄

**Comment by BethanyG** [comment]
@PaulT89   _**thank you so much for this!!**_  😄  🌟 

It might take me a bit tonight to review this.  Just to let you know how I typically review something this big (_please let me know if I need to adjust my strategy_):

1.  I make a pass to see if all the files are in place and things like configuration look good.
2.  Look at topic/subject content, and see if there are topics that need to be added or omitted.
3.  Look at the clarity/detail of explanations to see if they need adjusting.
4.  Look at the code examples to see that they are correct, and if additional ones are needed.
5.  Look at the exemplar and test files for logic, syntax, formatting, et. al.
6.  Check test cases for coverage.  Suggest parameters or cases as needed.
6.  Run the exemplar file against the test file to see if both "work" correctly.   

Because initial edits can sometimes result in a substantial reworking of things, I leave language usage, spelling, punctuation, and formatting details to a later cycle or pass.

**Comment by BethanyG** [comment]
All Done!  🎉
